essay/swar/latency.md

```diff
@@ -67,49 +67,114 @@ The essence of parallel algorithms is the "mapping" of an operation that is asso
 
-As with much of what superficially looks bad in our SWAR library, upon closer examination, we determine fairly rigorously that these are the best choices, the essential costs.  In that regard, it meets the primary mission of serving as a way to express in software all that we know about performance so that users can benefit the most from our expertise.
+As with much of what superficially may look bad in our SWAR library, upon closer examination, we determine fairly rigorously that these are the best choices, the essential costs.  In that regard, it meets the primary mission of serving as a way to express in software all that we know about performance so that users can benefit the most from our expertise
 
 
 -ed
 
 ## Footnotes
 
-[^1] We're beginning to work on explaining the basics of processor instructions and operations costs.
+[^1] We say "there are enough execution resources to consume 8 bytes of inputs at a time", we mean that multiplication of 64 bits seems to be as performant as 32 bits in most architectures we have used.  One possible exception might be GPUs: they might double the latency for 64 bit multiplications compared to 32.  We implemented Lemire's technique for 128 bits relatively unsuccessfully: 64 bit processors generally have instructions to help with 128 bits multiplies, for example, x86_64 gives you the result in the pair of registers `rDX:rAX` each of 64 bits.  We use this in the best 128 bit calque of the technique.  The straightforward use of full 128 bit multiplication requires three multiplications: two 64-to-64 bit, and one 64 bit factors to 128 bit result, this is better than the otherwise 2<sup>2</sup> (the usual multiplication algorihtm requires N<sup>2</sup> products), but most importantly, greater than 2, so, doubling the width performs worse because the critical operation takes 3 times as much, that's what we mean by having or not having execution resources for a given width:
+```c++
+auto multiply64_to128_bits(uint64_t f1, uint64_t f2) {
+    return __uint128_t(f1) * f2;
+}
+
+auto multiply128_to128_bits(__uint128_t f1, __uint128_t f2) {
+    return f1 * f2;
+}
+```
+translates to, via the [Compiler Explorer](https://godbolt.org/z/ofP71xzj9)
+```asm
+multiply64_to128_bits(unsigned long, unsigned long):            # @multiply64_to128_bits(unsigned long, unsigned long)
+        mov     rdx, rsi
+        mulx    rdx, rax, rdi
+        ret
+multiply128_to128_bits(unsigned __int128, unsigned __int128):           # @multiply128_to128_bits(unsigned __int128, unsigned __int128)
+        imul    rsi, rdx
+        mulx    rdx, rax, rdi
+        imul    rcx, rdi
+        add     rcx, rsi
+        add     rdx, rcx
+        ret
+```
+
+[^2] We're beginning to work on explaining the basics of processor instructions and operations costs.
+
+[^3] In AVX/AVX2 (of x86-64) typically a SIMD comparison such as "greater than" would generate a mask: for inputs `a` and `b`, the assembler would look like this (ChatGPT generated):
```

Comment:
Ugh, please, never mention the stochastic parrots by name. Best case they gave you something correct, worst case it is a crutch understood by your readers as a near verboten substitute for the sentence "I didn't understand what I published."